crates/core/src/buc/store/memory.rs

```diff
@@ -0,0 +1,268 @@
+use std::{
+	collections::BTreeMap,
+	future::Future,
+	pin::Pin,
+	sync::{Arc, RwLock},
+};
+
+use bytes::Bytes;
+use url::Url;
+
+use crate::sql::Datetime;
+
+use super::{Key, ListOptions, ObjectMeta, ObjectStore};
+
+#[derive(Clone, Debug, Default)]
+pub struct Entry {
+	bytes: Bytes,
+	updated: Datetime,
+}
+
+impl From<Bytes> for Entry {
+	fn from(bytes: Bytes) -> Self {
+		Self {
+			bytes,
+			..Default::default()
+		}
+	}
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct MemoryStore {
+	store: Arc<RwLock<BTreeMap<Key, Entry>>>,
+}
+
+impl MemoryStore {
+	pub fn new() -> Self {
+		MemoryStore::default()
+	}
+
+	pub fn parse_url(url: &str) -> bool {
+		if url == "memory" {
+			return true;
+		}
+
+		let Ok(url) = Url::parse(url) else {
+			return false;
+		};
+
+		url.scheme() == "memory"
+	}
+}
+
+impl ObjectStore for MemoryStore {
+	fn prefix(&self) -> Option<Key> {
+		None
+	}
+
+	fn put<'a>(
+		&'a self,
+		key: &'a Key,
+		data: Bytes,
+	) -> Pin<Box<dyn Future<Output = Result<(), String>> + Send + 'a>> {
+		Box::pin(async move {
+			let mut store =
+				self.store.write().map_err(|_| "Failed to lock object store".to_string())?;
+
+			store.insert(key.clone(), data.into());
+			Ok(())
+		})
+	}
+
+	fn put_if_not_exists<'a>(
+		&'a self,
+		key: &'a Key,
+		data: Bytes,
+	) -> Pin<Box<dyn Future<Output = Result<(), String>> + Send + 'a>> {
+		Box::pin(async move {
+			let mut store =
+				self.store.write().map_err(|_| "Failed to lock object store".to_string())?;
+
+			store.entry(key.clone()).or_insert_with(|| data.into());
+
+			Ok(())
+		})
+	}
+
+	fn get<'a>(
+		&'a self,
+		key: &'a Key,
+	) -> Pin<Box<dyn Future<Output = Result<Option<Bytes>, String>> + Send + 'a>> {
+		Box::pin(async move {
+			let store = self.store.read().map_err(|_| "Failed to lock object store".to_string())?;
+			let data = store.get(key).map(|v| v.bytes.clone());
+			Ok(data)
+		})
+	}
+
+	fn head<'a>(
+		&'a self,
+		key: &'a Key,
+	) -> Pin<Box<dyn Future<Output = Result<Option<ObjectMeta>, String>> + Send + 'a>> {
+		Box::pin(async move {
+			let store = self.store.read().map_err(|_| "Failed to lock object store".to_string())?;
+
+			let data = store.get(key).map(|v| ObjectMeta {
+				size: v.bytes.len() as u64,
+				updated: v.updated.clone(),
+				key: key.to_owned(),
+			});
+
+			Ok(data)
+		})
+	}
+
+	fn delete<'a>(
+		&'a self,
+		key: &'a Key,
+	) -> Pin<Box<dyn Future<Output = Result<(), String>> + Send + 'a>> {
+		Box::pin(async move {
+			let mut store =
+				self.store.write().map_err(|_| "Failed to lock object store".to_string())?;
```

Comment:
Probably worth putting the string "Failed to lock object store"  in a constant

Reply:
Converted to Dashmap in [120e55e](https://github.com/surrealdb/surrealdb/pull/5701/commits/120e55e0ad47bb91671a7d96931d22c5488e2b96), so no longer needed